ENH: Add  method to Image View so users can easily hook-up data processing code of their own.

hhslepicka opened the pull request:
MNT: Refactored the image processing example to work with the new callback.

I also took the opportunity to implement a stop and start on the timer when we reach and leave the redraw_image method. This will ensure that we don't try to process a new image while we have not yet finished with the previous one.

teddyrendahl commented on pydm/widgets/image.py:
```diff
@@ -287,18 +287,39 @@ def image_width_changed(self, new_width):
+    @staticmethod
+    def process_image(image):
+        """
+        Boilerplate method to be used by applications in order to add calculations and also modify the image before
```
Docstring should be one line 😄

anacso17 replied:
I'm not sure, but I think, by pep8, docstrings should have a short summary line (< 79 characters, ending with a period), followed by a blank line and then a description.

hhslepicka commented on examples/image_processing/image_view.py:
```diff
@@ -7,43 +7,36 @@
-        blobs= blob_doh(img, max_sigma=512, min_sigma=64, threshold=.02)
-        #Remove any existing blob markers        
+
+    def process_image(self, new_image):
```
@hmbui the reshape is inside the ImageView widget... it is more natural for users to think about the processing of a 2D image instead of a flat 1D array... so we only call process_image after reshaping.

hmbui replied:
👍

anacso17 commented on pydm/widgets/image.py:
```diff
@@ -287,18 +287,39 @@ def image_width_changed(self, new_width):
 
         If necessary, reshape the image to 2D first.
         """
+        self.redraw_timer.stop()
```
I think it is not necessary to stop/start `redraw_timer` within `redrawImage`, because QTimer will execute `redrawImage` again only after a current execution ends, what already ensures that we process a new image only after we finished with the previous one.

hhslepicka replied:
@anacso17 that is absolutely right! And this is also an issue... The QTimer is running the "processing" part in the `main thread` and this is nuking down the whole app when we add some calculation. What I am working now in a separated branch is to move the image reshape, processing and normalization to a QThread.
This will be better overall but will add a lot more of complexity to the user while writing their own code... that is why I opted to make `process_image` a static method... You can always get the ImageView widget from your code and act on the properties but you will not be allowed to touch directly QtWidgets from there...
I am updating now the `image_processing` example to see how it will come together.

anacso17 commented on pydm/widgets/image.py:
```diff
@@ -287,18 +287,39 @@ def image_width_changed(self, new_width):
             return
         self._image_width = int(new_width)
 
+    @staticmethod
```
Is it necessary to be a static method?
I have an application where I reimplemented `redrawImage` to add some processing, but I need some object attributes in this processing. So in this case, for example, I need this `process_image` to be a normal method.

hhslepicka replied:
Replied on the previous comment.

anacso17 commented on pydm/widgets/image.py:
```diff
@@ -287,18 +287,39 @@ def image_width_changed(self, new_width):
                 # We don't have a width for this image yet, so we can't draw it
+                self.redraw_timer.start()
                 return
             if self.readingOrder == ReadingOrder.Clike:
```
There is a bug here: `imageChannel` and `widthChannel` are independent, so `self.imageWidth` can be updated before some new `self.image_waveform` with this new width is received. It was a fix I was going to suggest.
In my application, where I reimplement `redrawImage`, I needed to put this `if self.readingOrder...` within a `try...except ValueError...` (where I just return to avoid breaking).

hhslepicka replied:
That makes sense for all control systems in which the variables are transmitted in a separated form. 😞 I will keep that in mind for the code that I am writing now.
PS: Would you mind to share a bit more of your current use-case?

anacso17 replied:
Ok! Thank you :)
Of course I can share. It is a work in progress, it was only tested with simulations. If you would like to take a look, it is available at https://github.com/lnls-sirius/hla/blob/master/pyqt-apps/siriushla/widgets/scrn_view.py.

anacso17 commented on pydm/widgets/image.py:
```diff
@@ -310,19 +331,22 @@ def redrawImage(self):
         else:
             mini = self.cm_min
             maxi = self.cm_max
+        img = self.process_image(img)
```
I think the image processing has to occur before `if self._normalize_data...`, because processing can change minimum and maximum values of the image.

hhslepicka replied:
I agree!

hhslepicka commented:
So here is what changed...
Now the steps to prepare an image are executed on a different Thread and with that, we no longer can poke the Widgets directly but instead we need to rely on Signals being emitted. 
It is annoying, sure but it is better than the screen freezing on users and poor performance.
Take the image_view or even image_processing examples for a spin and check how responsive the screen is now compared to before.
I will work on the camviewer later in the week. For now that is it...
Feel free to add more comments/concerns/ideas...
@anacso17 I addressed most if not all of your concerns here. Sorry about the mess that it will cause on your application but if you follow the guidelines of the image_processing example you will be fine. Let me know how it goes for you.

mattgibbs commented on pydm/widgets/image.py:
```diff
@@ -18,6 +22,53 @@ class ReadingOrder(object):
+        width = self.image_view.imageWidth
+        reading_order = self.image_view.readingOrder
+
+        self.image_view.lock.acquire()
```
This code could probably be made a little less brittle by using a `with <lock>` statement here, so that you don't have to make sure that you release the lock in every potential exit route.

mattgibbs commented on pydm/widgets/image.py:
```diff
@@ -287,11 +341,11 @@ def image_width_changed(self, new_width):
         """
-        Boilerplate method to be used by applications in order to add calculations and also modify the image before
-        it is displayed at the widget.
+        Boilerplate method to be used by applications in order to
```
Docs should probably warn that this gets called on a separate thread.

hhslepicka commented:
@teddyrendahl @mattgibbs @anacso17 this is good for another review.
Thank you all!

anacso17 commented:
Hey @hhslepicka, I tested the ImageViewer example a few times here and the window crashed after some time with the following error message:

           QThread: Destroyed while thread is still running
           Aborted (core dumped)

I think the crash was happening because at some point `self.lock.locked() = False` but `self.thread` was running yet. I think `threading.Lock` is not necessary for this implementation, once it is only used to verify if the ImageUpdateThread have finished. Here is a patch that solved the problem for me, without the lock:

```
diff --git a/pydm/widgets/image.py b/pydm/widgets/image.py
index 551d948..ceee08f 100644
--- a/pydm/widgets/image.py
+++ b/pydm/widgets/image.py
@@ -4,7 +4,6 @@ from pyqtgraph import ImageView
 from pyqtgraph import ColorMap
 from pyqtgraph.graphicsItems.ViewBox.ViewBoxMenu import ViewBoxMenu
 import numpy as np
-import threading
 import logging
 from .channel import PyDMChannel
 from .colormaps import cmaps, cmap_names, PyDMColorMap
@@ -39,30 +38,28 @@ class ImageUpdateThread(QThread):
         cm_min = self.image_view.cm_min
         cm_max = self.image_view.cm_max
 
-        with self.image_view.lock:
-            if not needs_redraw:
-                return
-            if image_dimensions == 1:
-                if width < 1:
-                    # We don't have a width for this image yet, so we can't draw it
-                    self.image_view.lock.release()
-                    return
-                if reading_order == ReadingOrder.Clike:
-                    img = img.reshape((-1, width), order='C')
-                else:
-                    img = img.reshape((width, -1), order='F')
-
-            if len(img) <= 0:
+        if not needs_redraw:
+            return
+        if image_dimensions == 1:
+            if width < 1:
+                # We don't have a width for this image yet, so we can't draw it
                 return
-            img = self.image_view.process_image(img)
-            if normalize_data:
-                mini = img.min()
-                maxi = img.max()
+            if reading_order == ReadingOrder.Clike:
+                img = img.reshape((-1, width), order='C')
             else:
-                mini = cm_min
-                maxi = cm_max
-            self.updateSignal.emit([mini, maxi, img])
-            self.image_view.needs_redraw = False
+                img = img.reshape((width, -1), order='F')
+
+        if len(img) <= 0:
+            return
+        img = self.image_view.process_image(img)
+        if normalize_data:
+            mini = img.min()
+            maxi = img.max()
+        else:
+            mini = cm_min
+            maxi = cm_max
+        self.updateSignal.emit([mini, maxi, img])
+        self.image_view.needs_redraw = False
 
 
 class PyDMImageView(ImageView, PyDMWidget, PyDMColorMap, ReadingOrder):
@@ -101,7 +98,6 @@ class PyDMImageView(ImageView, PyDMWidget, PyDMColorMap, ReadingOrder):
         """Initialize widget."""
         ImageView.__init__(self, parent)
         PyDMWidget.__init__(self)
-        self.lock = threading.Lock()
         self.thread = None
         self.axes = dict({'t': None, "x": 0, "y": 1, "c": None})
         self._imagechannel = image_channel
@@ -368,7 +364,7 @@ class PyDMImageView(ImageView, PyDMWidget, PyDMColorMap, ReadingOrder):
 
         If necessary, reshape the image to 2D first.
         """
-        if self.lock.locked():
+        if self.thread is not None and not self.thread.isFinished():
             logger.warning("Image processing has taken longer than the refresh rate.")
             return
         self.thread = ImageUpdateThread(self)
```

anacso17 commented on pydm/widgets/image.py:
```diff
@@ -287,42 +341,75 @@ def image_width_changed(self, new_width):
-        self.needs_redraw = False
+            autoDownsample=self.autoDownsample)
+
+
```
Blank line left over

anacso17 commented on pydm/widgets/image.py:
```diff
@@ -287,42 +341,75 @@ def image_width_changed(self, new_width):
         """
-        if not self.needs_redraw:
+        if self.lock.locked():
+            logger.warning("Image processing has taken longer than the refresh rate.")
```
This line is too long (>79 characters) (pep8)

anacso17 commented on pydm/widgets/image.py:
```diff
@@ -287,42 +341,75 @@ def image_width_changed(self, new_width):
+        self.thread.start()
+
+    @pyqtSlot(list)
+    def updateDisplay(self, data):
```
I think this method can be private since only the thread uses it.

teddyrendahl commented on examples/image_processing/image_view.py:
```diff
@@ -1,49 +1,48 @@
-        #Remove any existing blob markers        
+
+    def draw_markers(self, *args, **kwargs):
+        self.markers_lock.acquire()
```
Use the `lock` as `contextmanager here. 

https://docs.python.org/3/library/threading.html#using-locks-conditions-and-semaphores-in-the-with-statement

hhslepicka commented:
Thank you, @anacso17 and @teddyrendahl .
I addressed the comments. Now we have the `logging.DEBUG` at critical parts too.
Please let me know if you find other problems.

teddyrendahl commented:
👍

hhslepicka commented:
Talking offline with @anacso17...
No error is happening with the new fixes from the suggested patch.
I will merge this and if something happens we will open an issue and address it there.
